> [ERROR] Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.11.0:validate (default) on project qute-generator: File '/home/runner/work/quarkus/quarkus/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ValueResolverGenerator.java' has not been previously formatted.  Please format file and commit before running validation! -> [Help 1]

--- a/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java
+++ b/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java
@@ -79,6 +79,7 @@ import io.quarkus.qute.SectionHelper;
 import io.quarkus.qute.SectionHelperFactory;
 import io.quarkus.qute.Template;
 import io.quarkus.qute.TemplateException;
+import io.quarkus.qute.TemplateExtension;
 import io.quarkus.qute.TemplateInstance;
 import io.quarkus.qute.TemplateLocator;
 import io.quarkus.qute.Variant;
@@ -121,6 +122,7 @@ public class QuteProcessor {
     static final DotName COLLECTION = DotName.createSimple(Collection.class.getName());
 
     private static final String MATCH_NAME = "matchName";
+    private static final String PRIORITY = "priority";
 
     @BuildStep
     FeatureBuildItem feature() {
@@ -351,6 +353,7 @@ public class QuteProcessor {
             }
         }
 
+        // Method-level annotations
         for (Entry<MethodInfo, AnnotationInstance> entry : methods.entrySet()) {
             MethodInfo method = entry.getKey();
             ExtensionMethodGenerator.validate(method);
@@ -359,11 +362,18 @@ public class QuteProcessor {
                     method.declaringClass().name());
         }
 
+        // Class-level annotations
         for (Entry<ClassInfo, AnnotationInstance> entry : classes.entrySet()) {
             ClassInfo clazz = entry.getKey();
             for (MethodInfo method : clazz.methods()) {
                 if (!Modifier.isStatic(method.flags()) || method.returnType().kind() == org.jboss.jandex.Type.Kind.VOID
-                        || method.parameters().isEmpty() || Modifier.isPrivate(method.flags()) || methods.containsKey(method)) {
+                        || method.parameters().isEmpty() || Modifier.isPrivate(method.flags())
+                        || ValueResolverGenerator.isSynthetic(method.flags())) {
+                    // Filter out non-static, synthetic, private and void methods with no params
+                    continue;
+                }
+                if (methods.containsKey(method)) {
+                    // Skip methods annotated with @TemplateExtension - method-level annotation takes precedence
                     continue;
                 }
                 produceExtensionMethod(index, extensionMethods, method, entry.getValue());
@@ -375,6 +385,7 @@ public class QuteProcessor {
 
     private void produceExtensionMethod(IndexView index, BuildProducer<TemplateExtensionMethodBuildItem> extensionMethods,
             MethodInfo method, AnnotationInstance extensionAnnotation) {
+        // Analyze matchName and priority so that it could be used during validation 
         String matchName = null;
         AnnotationValue matchNameValue = extensionAnnotation.value(MATCH_NAME);
         if (matchNameValue != null) {
@@ -383,8 +394,13 @@ public class QuteProcessor {
         if (matchName == null) {
             matchName = method.name();
         }
+        int priority = TemplateExtension.DEFAULT_PRIORITY;
+        AnnotationValue priorityValue = extensionAnnotation.value(PRIORITY);
+        if (priorityValue != null) {
+            priority = priorityValue.asInt();
+        }
         extensionMethods.produce(new TemplateExtensionMethodBuildItem(method, matchName,
-                index.getClassByName(method.parameters().get(0).name())));
+                index.getClassByName(method.parameters().get(0).name()), priority));
     }
 
     @BuildStep
@@ -616,7 +632,8 @@ public class QuteProcessor {
 
         ExtensionMethodGenerator extensionMethodGenerator = new ExtensionMethodGenerator(classOutput);
         for (TemplateExtensionMethodBuildItem templateExtension : templateExtensionMethods) {
-            extensionMethodGenerator.generate(templateExtension.getMethod(), templateExtension.getMatchName());
+            extensionMethodGenerator.generate(templateExtension.getMethod(), templateExtension.getMatchName(),
+                    templateExtension.getPriority());
         }
         generatedTypes.addAll(extensionMethodGenerator.getGeneratedTypes());
 

--- a/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/TemplateExtensionMethodBuildItem.java
+++ b/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/TemplateExtensionMethodBuildItem.java
@@ -16,11 +16,17 @@ public final class TemplateExtensionMethodBuildItem extends MultiBuildItem {
     private final MethodInfo method;
     private final String matchName;
     private final ClassInfo matchClass;
+    private final int priority;
 
     public TemplateExtensionMethodBuildItem(MethodInfo method, String matchName, ClassInfo matchClass) {
+        this(method, matchName, matchClass, TemplateExtension.DEFAULT_PRIORITY);
+    }
+
+    public TemplateExtensionMethodBuildItem(MethodInfo method, String matchName, ClassInfo matchClass, int priority) {
         this.method = method;
         this.matchName = matchName;
         this.matchClass = matchClass;
+        this.priority = priority;
     }
 
     public MethodInfo getMethod() {
@@ -33,6 +39,10 @@ public final class TemplateExtensionMethodBuildItem extends MultiBuildItem {
 
     public ClassInfo getMatchClass() {
         return matchClass;
+    }
+
+    public int getPriority() {
+        return priority;
     }
 
     boolean matchesClass(ClassInfo clazz) {

--- a/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/Foo.java
+++ b/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/Foo.java
@@ -8,6 +8,11 @@ public class Foo {
 
     public Charlie charlie;
 
+    // ValueResolverGenerator.DEFAULT_PRIORITY
+    public String pong() {
+        return "foo";
+    }
+
     public Foo(String name, Long age) {
         this.name = name;
         this.age = age;

--- a/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/TemplateExtensionMethodsTest.java
+++ b/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/TemplateExtensionMethodsTest.java
@@ -26,13 +26,15 @@ public class TemplateExtensionMethodsTest {
     @RegisterExtension
     static final QuarkusUnitTest config = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
-                    .addClasses(Foo.class, Extensions.class)
+                    .addClasses(Foo.class, Extensions.class, PrioritizedExtensions.class)
                     .addAsResource(new StringAsset("{foo.name.toLower} {foo.name.ignored} {foo.callMe(1)} {foo.baz}"),
                             "templates/foo.txt")
                     .addAsResource(new StringAsset("{baz.setScale(2,roundingMode)}"),
                             "templates/baz.txt")
                     .addAsResource(new StringAsset("{anyInt.foo('bing')}"),
-                            "templates/any.txt"));
+                            "templates/any.txt")
+                    .addAsResource(new StringAsset("{foo.pong}::{foo.name}"),
+                            "templates/priority.txt"));
 
     @Inject
     Template foo;
@@ -72,6 +74,11 @@ public class TemplateExtensionMethodsTest {
 
     }
 
+    @Test
+    public void testPriority() {
+        assertEquals("bravo::baz", engine.getTemplate("priority").data("foo", new Foo("baz", 10l)).render());
+    }
+
     @TemplateExtension
     public static class Extensions {
 
@@ -100,6 +107,28 @@ public class TemplateExtensionMethodsTest {
         static String any(Integer val, String name, String info) {
             return val + "=" + info;
         }
+    }
+
+    @TemplateExtension(matchName = "pong")
+    public static class PrioritizedExtensions {
+
+        // default priority - class-level annotation
+        static String alpha(Foo foo) {
+            return "alpha";
+        }
+
+        // Explicit priority, higher than ValueResolverGenerator.DEFAULT_PRIORITY
+        @TemplateExtension(matchName = "pong", priority = 100)
+        static String bravo(Foo foo) {
+            return "bravo";
+        }
+
+        // default priority - method-level annotation
+        @TemplateExtension(matchName = "pong")
+        static String charlie(Foo foo) {
+            return "charlie";
+        }
+
     }
 
 }

--- a/independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateExtension.java
+++ b/independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateExtension.java
@@ -36,8 +36,31 @@ import java.lang.annotation.Target;
 @Target({ METHOD, TYPE })
 public @interface TemplateExtension {
 
-    static final String ANY = "*";
+    /**
+     * Constant value for {@link #matchName()} indicating that the method name should be used.
+     */
+    String METHOD_NAME = "<<method name>>";
 
-    String matchName() default "";
+    /**
+     * Constant value for {@link #matchName()} indicating that any name matches.
+     */
+    String ANY = "*";
+
+    /**
+     * Constant value for {{@link #priority()}.
+     */
+    int DEFAULT_PRIORITY = 5;
+
+    /**
+     * 
+     * @return the name is used to match the property name
+     */
+    String matchName() default METHOD_NAME;
+
+    /**
+     * 
+     * @return the priority used by the generated value resolver
+     */
+    int priority() default DEFAULT_PRIORITY;
 
 }

--- a/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ExtensionMethodGenerator.java
+++ b/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ExtensionMethodGenerator.java
@@ -45,6 +45,9 @@ public class ExtensionMethodGenerator {
 
     public static final String SUFFIX = "_Extension" + ValueResolverGenerator.SUFFIX;
 
+    private static final String MATCH_NAME = "matchName";
+    private static final String PRIORITY = "priority";
+
     private final Set<String> generatedTypes;
     private final ClassOutput classOutput;
 
@@ -69,25 +72,25 @@ public class ExtensionMethodGenerator {
         }
     }
 
-    public void generate(MethodInfo method, String matchName) {
+    public void generate(MethodInfo method, String matchName, Integer priority) {
 
         // Validate the method first
         validate(method);
+        ClassInfo declaringClass = method.declaringClass();
+        AnnotationInstance extensionAnnotation = method.annotation(TEMPLATE_EXTENSION);
 
-        if (matchName == null) {
-            AnnotationInstance extensionAnnotation = method.annotation(TEMPLATE_EXTENSION);
-            if (extensionAnnotation != null) {
-                AnnotationValue matchNameValue = extensionAnnotation.value("matchName");
-                if (matchNameValue != null) {
-                    matchName = matchNameValue.asString();
-                }
+        if (matchName == null && extensionAnnotation != null) {
+            // No explicit name defined, try annotation
+            AnnotationValue matchNameValue = extensionAnnotation.value(MATCH_NAME);
+            if (matchNameValue != null) {
+                matchName = matchNameValue.asString();
             }
         }
-
-        if (matchName == null) {
+        if (matchName == null || matchName.equals(TemplateExtension.METHOD_NAME)) {
             matchName = method.name();
-        } else if (matchName.equals(TemplateExtension.ANY)) {
-            // The second parameter must be a string
+        }
+        if (matchName.equals(TemplateExtension.ANY)) {
+            // Special constant used - the second parameter must be a string
             if (method.parameters().size() < 2 || !method.parameters().get(1).name().equals(STRING)) {
                 throw new IllegalStateException(
                         "Template extension method matching multiple names must declare at least two parameters and the second parameter must be string: "
@@ -95,7 +98,17 @@ public class ExtensionMethodGenerator {
             }
         }
 
-        ClassInfo declaringClass = method.declaringClass();
+        if (priority == null && extensionAnnotation != null) {
+            // No explicit priority set, try annotation
+            AnnotationValue priorityValue = extensionAnnotation.value(PRIORITY);
+            if (priorityValue != null) {
+                priority = priorityValue.asInt();
+            }
+        }
+        if (priority == null) {
+            priority = TemplateExtension.DEFAULT_PRIORITY;
+        }
+
         String baseName;
         if (declaringClass.enclosingClass() != null) {
             baseName = simpleName(declaringClass.enclosingClass()) + ValueResolverGenerator.NESTED_SEPARATOR
@@ -112,17 +125,17 @@ public class ExtensionMethodGenerator {
         ClassCreator valueResolver = ClassCreator.builder().classOutput(classOutput).className(generatedName)
                 .interfaces(ValueResolver.class).build();
 
-        implementGetPriority(valueResolver);
+        implementGetPriority(valueResolver, priority);
         implementAppliesTo(valueResolver, method, matchName);
         implementResolve(valueResolver, declaringClass, method, matchName);
 
         valueResolver.close();
     }
 
-    private void implementGetPriority(ClassCreator valueResolver) {
+    private void implementGetPriority(ClassCreator valueResolver, int priority) {
         MethodCreator getPriority = valueResolver.getMethodCreator("getPriority", int.class)
                 .setModifiers(ACC_PUBLIC);
-        getPriority.returnValue(getPriority.load(5));
+        getPriority.returnValue(getPriority.load(priority));
     }
 
     private void implementResolve(ClassCreator valueResolver, ClassInfo declaringClass, MethodInfo method, String matchName) {

--- a/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ValueResolverGenerator.java
+++ b/independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ValueResolverGenerator.java
@@ -74,6 +74,8 @@ public class ValueResolverGenerator {
     public static final String IGNORE = "ignore";
     public static final String PROPERTIES = "properties";
 
+    public static final int DEFAULT_PRIORITY = 10;
+
     private final Set<String> analyzedTypes;
     private final Set<String> generatedTypes;
     private final IndexView index;
@@ -159,7 +161,7 @@ public class ValueResolverGenerator {
     private void implementGetPriority(ClassCreator valueResolver) {
         MethodCreator getPriority = valueResolver.getMethodCreator("getPriority", int.class)
                 .setModifiers(ACC_PUBLIC);
-        getPriority.returnValue(getPriority.load(10));
+        getPriority.returnValue(getPriority.load(DEFAULT_PRIORITY));
     }
 
     private void implementResolve(ClassCreator valueResolver, String clazzName, ClassInfo clazz,

--- a/independent-projects/qute/generator/src/test/java/io/quarkus/qute/generator/SimpleGeneratorTest.java
+++ b/independent-projects/qute/generator/src/test/java/io/quarkus/qute/generator/SimpleGeneratorTest.java
@@ -53,7 +53,7 @@ public class SimpleGeneratorTest {
         MethodInfo extensionMethod = index.getClassByName(DotName.createSimple(MyService.class.getName())).method(
                 "getDummy", Type.create(myServiceClazz.name(), Kind.CLASS), PrimitiveType.INT,
                 Type.create(DotName.createSimple(String.class.getName()), Kind.CLASS));
-        extensionMethodGenerator.generate(extensionMethod, null);
+        extensionMethodGenerator.generate(extensionMethod, null, null);
         generatedTypes.addAll(extensionMethodGenerator.getGeneratedTypes());
     }
 

